sklearn/utils/__init__.py

```diff
@@ -592,15 +610,22 @@ def resample(*arrays, replace=True, n_samples=None, random_state=None, stratify=
 
     # convert sparse matrices to CSR for row-based indexing
     arrays = [a.tocsr() if issparse(a) else a for a in arrays]
-    resampled_arrays = [_safe_indexing(a, indices) for a in arrays]
+
+    if copy:
+        resampled_arrays = [_safe_indexing(a, indices) for a in arrays]
+    else:
+        for a in arrays:
+            a[:] = _safe_indexing(a, indices)
```

Comment:
Unfortunately I am pretty sure that `_safe_indexing(a, indices)`  still produces an intermediate allocation of the size of `a`.

I it is not visible in the memory profile because this way of profiling the memory measures the memory usage before and after line 618, but not during the execution of line 618.

To confirm this (if you do not trust me :) you can try to write a Python script that calls shuffle(X) on a larger X (e.g. 1GB) and run it with memory_profiler in "sampling" mode (also known as "time-based mode") with the `mprof run` command instead of the line-by-line tracing mode.

I think the only way to implement in-place shuffling memory efficiently would be to implement the Durstenfeld variant of the Fisher–Yates algorithm:

   https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#The_modern_algorithm

This has several drawbacks though:

- first it needs an explicit loop of size `n_samples` which might be slow to do in Python and would be much faster in Cython.
- however while implementing it Cython should be easily doable when all inputs are numpy arrays, this would be much more challenging for other input data structures (e.g. Python lists, pandas DataFrames/Series, and so on.
- finally inplace assignments for pairs of indices that would work for all the datastructures supported by `_safe_indexing` would require a similar utility function (e.g. `_safe_index_assign(X, indices, values)`) because index based assignment has no consistent API for all those libraries.

Reply:
Thank you for your comment.
I understand that the problem is complicated.

This issue seems difficult for me to solve, so I'm sorry but I would like to close this PR...

Reply:
no problem.

Reply:
There might be a way forward to use `numpy.random.shuffle` as explained in https://github.com/scikit-learn/scikit-learn/issues/7754#issuecomment-998070113 but I am not sure how complex it would be to get the details right.